tests/Doctrine/Tests/DBAL/Functional/MasterSlaveConnectionTest.php

```diff
@@ -139,4 +139,19 @@ public function testMasterSlaveConnectionCloseAndReconnect()
         $conn->connect('master');
         $this->assertTrue($conn->isConnectedToMaster());
     }
+
+    public function testKeepSlaveNoMasterOnExecuteQueryAfterWriteOperation()
+    {
+        $conn = $this->createMasterSlaveConnection(true);
+        $conn->insert('master_slave_table', array('test_int' => 30));
+
+        $this->assertTrue($conn->isConnectedToMaster());
+
+        $sql = "SELECT count(*) as num FROM master_slave_table";
```

Comment:
What about `executeQuery` done with an `INSERT`, for example?

Reply:
`INSERT INTO ... SELECT ...` is what I meant.

Also, many RDBMS allow arbitrary subqueries that are not necessarily just `SELECT` queries. This includes running stored procedures as part of a `SELECT`

Reply:
By default a slave connection used for `executeQuery`, until called any of `executeUpdate`,`beginTransaction` eg methods which switch a connection back to the master and keep it. So we will have the issue anyway.

This PR helps only switch a connection back to the slaves, so if we do queries like
`SELECT ...` (slave)
`UPDATE ...` (master)
`SELECT ...` (master, but why? should be slave again)
It's cool if you have async replication and want to avoid eventual consistency with monotonic queries, but it is not really useful if you want to do load balancing on read-only queries

Reply:
Why `executeUpdate()` uses  a slave connection instead of master?